Dsh/jigsaws (#32)
gSchool/ada-1-intro-to-ruby: 03-language-features/03-gems-jigsaw.md

## Tues 9/15 Gems Jigsaw

## Learning Goals
- Have a basic grasp of how to use:
  - [Awesome Print](https://rubygems.org/gems/awesome_print/versions/1.8.0)
  - [Pry](https://rubygems.org/gems/pry/versions/0.11.3)
  - [Faker](https://rubygems.org/gems/faker/versions/1.8.7)
  - [Chronic](https://github.com/mojombo/chronic)
  - [Money](http://rubymoney.github.io/money/)
  - [Terminal Table](https://github.com/tj/terminal-table)

## Jigsaw

> We will break into teams to investigate various gems.  Each team will create a **5 minute** presentation on their assigned gem(s).  In this activity, we will be learning and then teaching each other, in the format of two rounds: the research round and the teaching round.

Reviewer: We say "gem" here, but say "Gem" later.

Author: eh.

1. [Money](http://rubymoney.github.io/money/)
1. [Terminal Table](https://github.com/tj/terminal-table)

In the teaching round, your research round group will disseminate and you will go into your a second assigned group. It is your responsibility to teach this group about that topic.

### What to Research
You will be learning and teaching each other a variety of Ruby Gems.

You will be practicing and teaching each other how to read through the Ruby documentation & Github pages to find these answers.

### Where to Start
You will want to start with the sample code below, and github pages for your given gem. To find the right web page, you may want to directly use a search engine, or you can look through the RubyGems docs for a link to the Gem. You will want to answer the questions:

-   What does this Gem do on a high level?
-   How do you use it?  Provide an example
-   What kind of syntax does it require?
-   Provide an example of a situation where it would be useful.

### Awesome Print

Awesome Print is... truly awesome!  It highlights syntax and gives you formatting options on console output.

```ruby
require "awesome_print"
```

**Example: Using the `awesome_print` gem.**

> Generally the github page for a gem is the best place to go to get started. Let's check out the [awesome_print gems' page](https://github.com/awesome-print/awesome_print). Under _usage_ we see some common examples of how to use this gem.

Reviewer: ```suggestion
Generally the github page for a gem is the best place to go to get started. Let's check out the [awesome_print github page](https://github.com/awesome-print/awesome_print). Under _usage_ we see some common examples of how to use this gem.
```

# We will stick to the basics.

# Comparing ap to puts

sample_array = ["Mercury", "Venus", "Earth", "Mars"]
ap "Awesome Print with an Array"
ap sample_array
puts "Puts with an Array"
puts sample_array

nested_structure = {
  points: [
    {
      name: "Great Pyramid of Giza",
      location: [32.5, 15.2]
    },
    {
      name: "Temple of Artemis",
      location: [47.8, 13.4]
    },
    {
      name: "Lighthouse of Alexandria",
      location: [32.6, 15.7]
    }
  ]
}

ap "Awesome Print with a nested structure"
ap nested_structure
puts "puts with a nested structure"
puts nested_structure

ap "Awesome print with an object"
class City
  def initialize(name)
    @name = name
  end
end

my_city = City.new("Seattle")
ap "Awesome Print with an Instance"
ap my_city
ap "Puts with an instance"
puts my_city

ap "Awesome print with Options!", color: {string: :purpleish}
```
![Awsome Print in Action!](images/awesome_print2.png)

**Exercise** Try to add an option to an `ap` statement.  Check the [github page](https://github.com/awesome-print/awesome_print)  for instructions

### Pry

Pry is an amazing gem which serves a couple of purposes.

> 1.  It can replace irb to serve as a REPL.  This is advantageous because it formats code nicely.

Reviewer: ```suggestion
1. It can replace `irb` to serve as a [REPL](https://en.wikipedia.org/wiki/Read%E2%80%93eval%E2%80%93print_loop).  This is advantageous because it formats code nicely.
```

To install `pry`
```bash
gem install pry
```

You can use `pry` to halt your program and open up a REPL to explore your running program with the command `binding.pry`

For example:

```ruby
# test_pry.rb
require "pry"

num = 0
while num < 10
  num += 1
  binding.pry
end

```

Running the program above with `ruby test_pry.rb` will result in this:
```
ruby test_pry.rb

From: /Users/chris/ada/ruby/test_pry.rb @ line 6 :

    1: require "pry"
    2:
    3: num = 0
    4: while num < 10
    5:   num += 1
 => 6:   binding.pry
    7: end

[1] pry(main)>
```

Pry has halted the program mid-execution and put you into an environment where you can interact with the program live.  You can then enter irb-like commands like: `num` to get the current value of the variable in the program, or call methods etc.

You can enter `exit` to exit the current binding and the program will halt again the next time it encounters  `binding.pry`, which in this case is the next iteration of the loop.  When you finish you can exit the entire program with `exit-program`.  We will go further into `pry` at a later date and you are welcome to [research](http://pryrepl.org/) the gem.

#### Replacing irb

> You can have Pry replace irb by editing your bash profile with `code ~/.profile` and adding the following line to the end of the file:

Reviewer: ```suggestion
You can have Pry replace irb by editing your zsh configuration with `mine ~/.zshrc` and adding the following line to the end of the file:
```
